src/python/pants/backend/docker/target_types.py

```diff
@@ -30,8 +33,24 @@ class DockerBuildArgsField(StringSequenceField):
 
 class DockerImageSourceField(SingleSourceField):
     default = "Dockerfile"
+    default_glob_match_error_behavior = GlobMatchErrorBehavior.ignore
+    expected_num_files = range(0, 2)
+    required = False
+    help = (
+        "The Dockerfile to use when building the Docker image.\n\n"
+        "Use the `instructions` field instead if you prefer not having the Dockerfile in your "
+        "project source tree."
```

Comment:
I was going to suggest implementing target validation with `Target.validate()` to check that exactly one of these two fields is set. Although, I remember now that isn't possible and is the whole motivation for your `source` `default` prework...

Still, maybe mention this expectation in the `help` for `docker_image` target in a new paragraph. And to be precise, you don't need to _define_ two fields thanks to the `source` default. Sounds like you either need to set `instructions` or have a file on-disk matching the `source` field, which defaults to `Dockerfile`